### PR TITLE
fix: reset character count on focus

### PR DESCRIPTION
**Pull request by szilvihorvath**
Issue: https://github.com/Capgemini/dcx-react-library/issues/437

Character count is reset on focus. So after the textarea has been cleared programmatically, and then the textarea is focussed, the counter and the message will be reset to the correct value.

_This is not the perfect solution as the count and message is not updated on clicking the button that clears the form._

### CODE REVIEW

File: src/characterCount/CharacterCount.tsx [hunk 1]
```diff
@@ -107,119 +116,163 @@ type CharacterCountProps = React.HTMLAttributes<HTMLTextAreaElement> & {
     overLimitBy?: number,
     hydrated?: boolean
   ) => string;
+  /**
+   * will allow to expose the reset function to clear textbox and reset the component;
+   */
+  ref?: any;
```
Comment (daniele-zurico) [inline src/characterCount/CharacterCount.tsx:122]: can we use a better type?

File: src/characterCount/CharacterCount.tsx [hunk 2]
```diff
@@ -107,119 +116,163 @@ type CharacterCountProps = React.HTMLAttributes<HTMLTextAreaElement> & {
     overLimitBy?: number,
     hydrated?: boolean
   ) => string;
+  /**
+   * will allow to expose the reset function to clear textbox and reset the component;
```
Comment (daniele-zurico) [inline src/characterCount/CharacterCount.tsx:120]: can we add below something like:
```
@Capgemini/tdd-examples-devs 
const textRef = useRef<any>(null);
<Button onClick={() => textRef.current.reset()} label="Cancel" />
```
this should appear in storybook documentation

File: src/characterCount/CharacterCount.tsx [hunk 3]
```diff
@@ -107,119 +116,163 @@ type CharacterCountProps = React.HTMLAttributes<HTMLTextAreaElement> & {
...
+      setTextareaValue(value);
+    }, [value]);
+
+    useEffect(() => {
+      const remaining = getRemaining(textareaValue);
+      const overThreshold = isOverThreshold(textareaValue);
+
+      setShowMessage(overThreshold || !threshold);
+      if (remaining < 0) setOverLimitBy(-remaining);
+      setRemainingCount(remaining);
```
Comment (daniele-zurico) [inline src/characterCount/CharacterCount.tsx:192]: I'll suggest to extract a function and call it here... I suppose what is doing here is to calculate the remaining count so something like:
```js
const calcRemainingCount = () => {
const remaining = getRemaining(textareaValue);
      const overThreshold = isOverThreshold(textareaValue);

      setShowMessage(overThreshold || !threshold);
      if (remaining < 0) setOverLimitBy(-remaining);
      setRemainingCount(remaining);
}

useEffect(()=> calcRemainingCount, [textareaValue]);
```

File: src/characterCount/CharacterCount.tsx [hunk 4]
```diff
@@ -107,119 +116,163 @@ type CharacterCountProps = React.HTMLAttributes<HTMLTextAreaElement> & {
...
+            onChange={handleChange}
+            rows={rows}
+            cols={cols}
+            maxLength={constrained && !isWordsCount ? maxLength : undefined}
+            aria-describedby={ariaDescribedBy || ''}
+            onFocus={handleChange}
+            onReset={handleChange}
+            ref={ref}
+            {...props}
+            value={textareaValue}
```
Comment (daniele-zurico) [inline src/characterCount/CharacterCount.tsx:258]: swap the order:
```html
value={textareaValue}
{...props}
```
spreading the operator should always go at the end in this way will override whatever has been defined before

**Review by daniele-zurico — APPROVED** [review]

**Review by Ibabalola — APPROVED** [review]

**Pull request merged**